File: backend/internal/permissions/permissions.go
package permissions

// Permission flags using bitwise operations
const (
    // Project Permissions
    PermViewAllProjects    uint32 = 1 << iota  // Admin: view all projects (and companies?)
    PermReviewProjects                         // Admin: review/approve/decline projects
    PermManageUsers                            // Admin: manage user accounts
    PermManagePermissions                      // Admin: modify user permissions
    PermSubmitProject                          // Startup: submit projects for review
    PermCommentOnProjects                      // Investor: comment on projects
    PermInvestInProjects                       // Investor: invest in projects
    PermManageDocuments                        // Startup: manage project documents
    PermManageInvestments                      // Admin: manage investments
    PermManageTeam                             // Startup: manage team members
    PermIsAdmin                                // Special bit to identify admins
    
    // This will be 2^(number of permissions) - 1, making a mask of all valid bits (POG)
    validPermissionBits = (1 << (iota)) - 1
)

// Role-based permission sets
const (
    PermAdmin = PermIsAdmin | PermViewAllProjects | PermReviewProjects | 
                PermManageUsers | PermManagePermissions | PermCommentOnProjects

    PermStartupOwner = PermSubmitProject | PermCommentOnProjects | PermManageDocuments | PermManageTeam

    PermInvestor = PermViewAllProjects | PermCommentOnProjects | PermInvestInProjects

    PermRegular uint32 = 0 // This is a placeholder for the new role
)

// GetAllPermissionBits returns a mask containing all valid permission bits.
// This is used for validation to ensure no invalid bits are set.
func GetAllPermissionBits() uint32 {
    return validPermissionBits
}

// Helper functions for permission checking
func HasPermission(userPerms, requiredPerm uint32) bool {
    return userPerms&requiredPerm != 0
}

func HasAllPermissions(userPerms uint32, requiredPerms ...uint32) bool {
    for _, perm := range requiredPerms {
        if !HasPermission(userPerms, perm) {
            return false
        }
    }
    return true
}

func HasAnyPermission(userPerms uint32, requiredPerms ...uint32) bool {
    for _, perm := range requiredPerms {
        if HasPermission(userPerms, perm) {
            return true
        }
    }
    return false
} 

File: frontend/src/utils/permissions.ts
/**
 * Permission flags matching backend permissions
 */
export const Permissions = {
    // Project Permissions
    PermViewAllProjects: 1 << 0, // Admin: view all projects (and companies?)
    PermReviewProjects: 1 << 1, // Admin: review/approve/decline projects
    PermManageUsers: 1 << 2, // Admin: manage user accounts
    PermManagePermissions: 1 << 3, // Admin: modify user permissions
    PermSubmitProject: 1 << 4, // Startup: submit projects for review
    PermCommentOnProjects: 1 << 5, // All: comment on projects
    PermInvestInProjects: 1 << 6, // Investor: invest in projects
    PermManageDocuments: 1 << 7, // Startup: manage project documents
    PermManageInvestments: 1 << 8, // Admin: manage investments
    PermManageTeam: 1 << 9, // Startup: manage team members
    PermIsAdmin: 1 << 10, // Special bit to identify admins
} as const;

/**
 * Checks if the user has all the specified permissions
 */
export function hasAllPermissions(
    userPermissions: number,
    ...requiredPermissions: number[]
): boolean {
    // Combine all required permissions into a single mask
    const requiredMask = requiredPermissions.reduce(
        (mask, perm) => mask | perm,
        0
    );
    // Check if all required bits are set in userPermissions
    return (userPermissions & requiredMask) === requiredMask;
}

/**
 * Checks if the user has any of the specified permissions
 */
export function hasAnyPermission(
    userPermissions: number,
    ...requiredPermissions: number[]
): boolean {
    return requiredPermissions.some(
        (permission) => (userPermissions & permission) !== 0
    );
}

/**
 * Checks if the user has admin permissions
 */
export function isAdmin(userPermissions: number): boolean {
    // For admin, we just need to check the admin bit lol
    return (
        (userPermissions & Permissions.PermIsAdmin) === Permissions.PermIsAdmin
    );
}

/**
 * Checks if the user has startup owner permissions
 */
export function isStartupOwner(userPermissions: number): boolean {
    return hasAllPermissions(
        userPermissions,
        Permissions.PermSubmitProject,
        Permissions.PermManageDocuments,
        Permissions.PermManageTeam,
        Permissions.PermCommentOnProjects
    );
}

/**
 * Checks if the user has investor permissions
 */
export function isInvestor(userPermissions: number): boolean {
    return hasAllPermissions(
        userPermissions,
        Permissions.PermViewAllProjects,
        Permissions.PermCommentOnProjects,
        Permissions.PermInvestInProjects
    );
}

/**
 * Gets an array of permission names that a user has
 */
export function getPermissionNames(userPermissions: number): string[] {
    return Object.entries(Permissions)
        .filter(([_, value]) => (userPermissions & value) === value)
        .map(([key]) => key);
}
